src/main/scala/cloud/drdrdr/utils/AWSCredentials.scala

```diff
@@ -6,165 +6,416 @@ package cloud.drdrdr.utils
 
 
 import java.io.File
+import java.text.SimpleDateFormat
+import java.util.{Date, SimpleTimeZone}
 
 import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
 import akka.http.scaladsl.model._
-import akka.stream.ActorMaterializer
-import akka.stream.scaladsl.Sink
 import akka.pattern.after
-import cloud.drdrdr.SignRequestForAWS
-import scala.io.Source
+import akka.stream.{ActorMaterializer, scaladsl}
+import akka.stream.scaladsl.Sink
+import akka.util.ByteString
 import spray.json._
+import org.slf4j.LoggerFactory
+import com.typesafe.scalalogging.Logger
 import scala.concurrent.duration._
-import scala.concurrent.{Promise, Await, ExecutionContext, Future}
-import scala.util.{Success, Failure}
+import scala.concurrent.{ExecutionContext, Future, Await}
+import scala.io.Source
+import scala.util.control
+
+
+trait AWSCredentials {
+
+  case class AWSCredentialSource(creds: Future[AWSPermissions]) (implicit ec: ExecutionContext, s:ActorSystem, m: ActorMaterializer) {
+    private var credentials: Future[AWSPermissions] = creds
+
+    /**
+     * Updates the credentials if necessary, then returns the credentials
+     * @return AWSPermissions with most recent credentials
+     */
+    def getCredentials: Future[AWSPermissions] = {
+      credentials flatMap {
+        case perm:AWSPermissions =>
```

Comment:
I think this case is non-exhaustive, and maybe its not necessary to check the type because its set in stone up in line 31. Could be simplified by doing `credentials flatMap { perm => `